apm-agent-plugins/apm-jms-plugin/src/test/java/co/elastic/apm/agent/jms/JmsInstrumentationIT.java

```diff
@@ -112,87 +123,211 @@ public void endTransaction() throws Exception {
+        if (!receiveNoWaitFlow.get()) {
+            assertThat(sendToNoopSpans).hasSize(1);
+            sendToNoopSpan = sendToNoopSpans.get(0);
+        }
+
+        // rest of spans should be receive spans yielding null messages
+        final String receiveWithNoMessageSpanName = "JMS RECEIVE";
+        assertThat(spans.stream().filter(span -> span.getNameAsString().equals(receiveWithNoMessageSpanName)).count()).isGreaterThanOrEqualTo(numSpans - 3);
 
         //noinspection ConstantConditions
         Id currentTraceId = tracer.currentTransaction().getTraceContext().getTraceId();
         assertThat(sendSpan.getTraceContext().getTraceId()).isEqualTo(currentTraceId);
 
-        Transaction receiveTransaction = reporter.getFirstTransaction();
-        assertThat(receiveTransaction.getTraceContext().getTraceId()).isNotEqualTo(currentTraceId);
-        assertThat(receiveSpan.getTraceContext().getTraceId()).isEqualTo(receiveTransaction.getTraceContext().getTraceId());
+        Id receiveTraceId = receiveSpan.getTraceContext().getTraceId();
+        List<Transaction> receiveTransactions = reporter.getTransactions().stream().filter(transaction -> transaction.getTraceContext().getTraceId().equals(receiveTraceId)).collect(Collectors.toList());
+        assertThat(receiveTransactions).hasSize(1);
+        Transaction receiveTransaction = receiveTransactions.get(0);
         assertThat(receiveSpan.getTraceContext().getParentId()).isEqualTo(receiveTransaction.getTraceContext().getId());
 
+        if (sendToNoopSpan != null) {
+            assertThat(sendToNoopSpan.getTraceContext().getTraceId()).isEqualTo(receiveTraceId);
+            assertThat(sendToNoopSpan.getTraceContext().getParentId()).isEqualTo(receiveTransaction.getTraceContext().getId());
+        }
     }
 
-    private void testQueueSendReceiveOnNonTracedThread(Callable<Message> receiveMethod, Queue queue, boolean waitAfterSend)
+    // tests transaction creation following a receive
+    private void verifyQueueSendReceiveOnNonTracedThread(Queue queue)
         throws Exception {
         String message = UUID.randomUUID().toString();
         Message outgoingMessage = brokerFacade.createTextMessage(message);
         brokerFacade.send(queue, outgoingMessage);
-        if (waitAfterSend) {
-            Thread.sleep(100);
-        }
-        Message incomingMessage = receiveOnNonTracedThread(receiveMethod).get(3, TimeUnit.SECONDS);
+        Message incomingMessage = resultQ.poll(2, TimeUnit.SECONDS);
+        assertThat(incomingMessage).isNotNull();
         verifyMessage(message, incomingMessage);
-        verifySendReceiveOnNonTracedThread(queue.getQueueName(), 1);
+        verifySendReceiveOnNonTracedThread(queue.getQueueName());
     }
 
-    private void verifySendReceiveOnNonTracedThread(String destinationName, int expectedReadTransactions) {
+    private void verifySendListenOnNonTracedThread(String destinationName, int expectedReadTransactions) {
```

Comment:
Could we use [awaitility](https://github.com/awaitility/awaitility) in order to get rid of some `Thread.sleep`s? 

```java
await().atMost(500, MILLISECONDS).until(() -> !reporter.getSpans().isEmpty());
```

Sleeps are not only a source of flakiness but also make it hard to follow on which condition we are waiting for.